File: C++/DesignPatterns/RATELIMITING.cpp
#include <algorithm>
#include <chrono>
#include <deque>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>

static double nowSeconds() {
    using namespace std::chrono;
    return duration<double>(steady_clock::now().time_since_epoch()).count();
}

struct Request {
    std::string clientId;
    double timestamp;

    explicit Request(std::string id, double ts = nowSeconds())
        : clientId(std::move(id)), timestamp(ts) {}
};

struct IRateLimitingStrategy {
    virtual ~IRateLimitingStrategy() = default;
    virtual bool isAllowed(const Request& request) = 0;
};

class TokenBucketStrategy : public IRateLimitingStrategy {
public:
    TokenBucketStrategy(int capacity, double refillRatePerSec)
        : capacity_(capacity),
          refillRatePerSec_(refillRatePerSec),
          tokens_(capacity),
          lastRefill_(nowSeconds()) {}

    bool isAllowed(const Request& request) override {
        refill(request.timestamp);
        if (tokens_ >= 1.0) {
            tokens_ -= 1.0;
            return true;
        }
        return false;
    }

private:
    void refill(double now) {
        double elapsed = now - lastRefill_;
        tokens_ = std::min(static_cast<double>(capacity_),
                           tokens_ + elapsed * refillRatePerSec_);
        lastRefill_ = now;
    }

    int capacity_;
    double refillRatePerSec_;
    double tokens_;
    double lastRefill_;
};

class SlidingWindowStrategy : public IRateLimitingStrategy {
public:
    SlidingWindowStrategy(double windowSizeSec, int maxRequests)
        : windowSizeSec_(windowSizeSec), maxRequests_(maxRequests) {}

    bool isAllowed(const Request& request) override {
        double now = request.timestamp;
        while (!timestamps_.empty() && timestamps_.front() <= now - windowSizeSec_) {
            timestamps_.pop_front();
        }
        if (static_cast<int>(timestamps_.size()) < maxRequests_) {
            timestamps_.push_back(now);
            return true;
        }
        return false;
    }

private:
    double windowSizeSec_;
    int maxRequests_;
    std::deque<double> timestamps_;
};

class NotificationService {
public:
    void setStrategy(std::unique_ptr<IRateLimitingStrategy> strategy) {
        strategy_ = std::move(strategy);
    }

    void send(const Request& request) {
        if (!strategy_) {
            throw std::runtime_error("No rate limiting strategy configured");
        }
        if (strategy_->isAllowed(request)) {
            std::cout << "Notification sent for client " << request.clientId << std::endl;
        } else {
            std::cout << "Rate limit exceeded for client " << request.clientId << std::endl;
        }
    }

private:
    std::unique_ptr<IRateLimitingStrategy> strategy_;
};

int main() {
    NotificationService service;

    service.setStrategy(std::make_unique<TokenBucketStrategy>(3, 1.0));
    std::cout << "-- Token Bucket --" << std::endl;
    for (int i = 0; i < 5; ++i) {
        service.send(Request("user-1"));
    }

    service.setStrategy(std::make_unique<SlidingWindowStrategy>(60.0, 2));
    std::cout << "-- Sliding Window --" << std::endl;
    for (int i = 0; i < 4; ++i) {
        service.send(Request("user-2"));
    }

    return 0;
}

File: Python/DesignPatterns/RATELIMITING.py
from abc import ABC, abstractmethod
from collections import deque
import time


class Request:
    def __init__(self, client_id: str, timestamp: float | None = None) -> None:
        self.client_id = client_id
        self.timestamp = timestamp if timestamp is not None else time.time()


class IRateLimitingStrategy(ABC):
    @abstractmethod
    def is_allowed(self, request: Request) -> bool:
        ...


class TokenBucketStrategy(IRateLimitingStrategy):
    def __init__(self, capacity: int, refill_rate_per_sec: float) -> None:
        self.capacity = capacity
        self.refill_rate_per_sec = refill_rate_per_sec
        self.tokens: float = capacity
        self.last_refill = time.time()

    def _refill(self, now: float) -> None:
        elapsed = now - self.last_refill
        self.tokens = min(self.capacity, self.tokens + elapsed * self.refill_rate_per_sec)
        self.last_refill = now

    def is_allowed(self, request: Request) -> bool:
        self._refill(request.timestamp)
        if self.tokens >= 1:
            self.tokens -= 1
            return True
        return False


class SlidingWindowStrategy(IRateLimitingStrategy):
    def __init__(self, window_size_sec: float, max_requests: int) -> None:
        self.window_size_sec = window_size_sec
        self.max_requests = max_requests
        self.timestamps: deque[float] = deque()

    def is_allowed(self, request: Request) -> bool:
        now = request.timestamp
        while self.timestamps and self.timestamps[0] <= now - self.window_size_sec:
            self.timestamps.popleft()
        if len(self.timestamps) < self.max_requests:
            self.timestamps.append(now)
            return True
        return False


class NotificationService:
    def __init__(self) -> None:
        self.strategy: IRateLimitingStrategy | None = None

    def set_strategy(self, strategy: IRateLimitingStrategy) -> None:
        self.strategy = strategy

    def send(self, request: Request) -> None:
        if self.strategy is None:
            raise RuntimeError("No rate limiting strategy configured")
        if self.strategy.is_allowed(request):
            print(f"Notification sent for client {request.client_id}")
        else:
            print(f"Rate limit exceeded for client {request.client_id}")


def main() -> None:
    service = NotificationService()

    service.set_strategy(TokenBucketStrategy(capacity=3, refill_rate_per_sec=1))
    print("-- Token Bucket --")
    for _ in range(5):
        service.send(Request("user-1"))

    service.set_strategy(SlidingWindowStrategy(window_size_sec=60, max_requests=2))
    print("-- Sliding Window --")
    for _ in range(4):
        service.send(Request("user-2"))


if __name__ == "__main__":
    main()
